src/ripple/app/misc/impl/AMM.cpp

```diff
@@ -0,0 +1,276 @@
+    Copyright (c) 2022 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+#include <ripple/app/misc/AMM.h>
+#include <ripple/app/paths/TrustLine.h>
+#include <ripple/ledger/Sandbox.h>
+#include <ripple/protocol/STArray.h>
+
+namespace ripple {
+
+uint256
+calcAMMGroupHash(Issue const& issue1, Issue const& issue2)
+{
+    if (issue1 < issue2)
+        return sha512Half(
+            issue1.account, issue1.currency, issue2.account, issue2.currency);
+    return sha512Half(
+        issue2.account, issue2.currency, issue1.account, issue1.currency);
+}
+
+Currency
+calcLPTCurrency(AccountID const& ammAccountID)
+{
+    return Currency::fromVoid(ammAccountID.data());
+}
+
+Issue
+calcLPTIssue(AccountID const& ammAccountID)
+{
+    return Issue(calcLPTCurrency(ammAccountID), ammAccountID);
+}
+
+std::pair<STAmount, STAmount>
+ammPoolHolds(
+    ReadView const& view,
+    AccountID const& ammAccountID,
+    Issue const& issue1,
+    Issue const& issue2,
+    beast::Journal const j)
+{
+    auto const assetInBalance = accountHolds(
+        view,
+        ammAccountID,
+        issue1.currency,
+        issue1.account,
+        FreezeHandling::fhZERO_IF_FROZEN,
+        j);
+    auto const assetOutBalance = accountHolds(
+        view,
+        ammAccountID,
+        issue2.currency,
+        issue2.account,
+        FreezeHandling::fhZERO_IF_FROZEN,
+        j);
+    return std::make_pair(assetInBalance, assetOutBalance);
+}
+
+std::tuple<STAmount, STAmount, STAmount>
+ammHolds(
+    ReadView const& view,
+    SLE const& ammSle,
+    std::optional<Issue> const& optIssue1,
+    std::optional<Issue> const& optIssue2,
+    beast::Journal const j)
+{
+    auto const [issue1, issue2] = [&]() -> std::pair<Issue, Issue> {
+        if (optIssue1 && optIssue2)
+            return {*optIssue1, *optIssue2};
+        auto const [issue1, issue2] = getTokensIssue(ammSle);
+        if (optIssue1)
+        {
+            if (*optIssue1 == issue1)
+                return {issue1, issue2};
+            else if (*optIssue1 == issue2)
+                return {issue2, issue1};
+            Throw<std::runtime_error>("ammHolds: Invalid optIssue1.");
```

Comment:
If this function throws then apply will return `tefException` - which is not correct. That would imply some sort of logic error in the code, but this exception can be thrown because the user provided incorrect transaction data.

I'd suggest changing all the functions that can currently throw on user input to use `Expected` instead and then we can insure that the transactions return the correct error codes. We could also try to wrap all the call with try/catch blocks, but that is more likely to miss cases where we should have caught the exception. `Expected` makes that much less likely (at the cost of more awkward code). I'd prefer `Expected` for this patch.

Reply:
Got it. Will change.